http/peer-id-auth.md

```diff
@@ -0,0 +1,194 @@
+
+Params are encoded per [RFC 9110 auth-param's ABNF](https://datatracker.ietf.org/doc/html/rfc9110#name-collected-abnf). Generally it'll be something like: `hostname="example.com", challenge-server=base64EncodedVal`
+
+## Signing
+
+Signatures sign some set of parameters. The parameters are sorted
+alphabetically, prepended with a varint length prefix, and concatenated together
+to form the data to be signed. The signing algorithm is defined by the key type
+used. Refer to the [Peer ID
+spec] for
+specifics on the signing algorithm. The set of parameters is prefixed with the auth scheme "libp2p-PeerID"
+
+As an example, if we wanted to sign the parameters `hostname="example.com",
+challenge-client=base64String` we would first structure the parameters as a byte
+slice containing:
+```
+libp2p-PeerID<varintprefix>challenge-client=<base64String><varintprefix>hostname="example.com"
+```
+
+Then sign the resulting byte slice. See the test vectors below for a
+examples.
+
+
+## Base64 Encoding
+
+The base64 encoding follows Base 64 Encoding with URL and Filename Safe Alphabet
+from [RFC 4648](https://datatracker.ietf.org/doc/html/rfc4648#section-5). The
+reason this is not a multibase is to aid clients or servers who can not or
+prefer not to import a multibase dependency.
+
+## Mutual Client and Server Peer ID
+
+The following protocol allows both the client and server to authenticate each
+other's peer id by having them each sign a challenge issued by the other. The
+protocol operates as follows:
+
+1. The client makes a POST request to the authentication URI.
+2. The server responds with status code 401 (Unauthorized) and set the header:
+   ```
+   WWW-Authenticate: libp2p-PeerID challenge-client=<base64-encoded-challenge>, opaque=<base64-encoded-opaque-value>
+   ```
+   The opaque parameter is opaque to client. The client MUST return the opaque
+   parameter back to the server. The server MAY use the opaque parameter to
+   encode state.
+3. The client makes another POST request to the authentication URI and sets the header:
+   ```
+   Authorization: libp2p-PeerID peer-id=<string-representation-of-client-peer-id>, opaque=<opaque-from-server>, challenge-server=<base64-encoded-challenge-server>, sig=<base64-signature-bytes>
+   ```
+
+   The `sig` param represents a signature over the parameters:
+    - `hostname`
+    - `challenge-client` in its base64 encoded form.
+4. The server MUST verify the signature using the server name used in the TLS
+   session. The server MUST return 401 Unauthorized if the server fails to
+   validate the signature. If the signature is valid, the server has
+   authenticated the client's peer id. The server MUST respond with status code
+   200 (OK) and set the headers:
+   ```
+   Authentication-Info: libp2p-PeerID peer-id=<server-peer-id-string>, sig=<base64-signature-bytes>
+   Authorization: libp2p-Bearer <base64-encoded-opaque-blob>
```

Comment:
Is Authorization generally used with HTTP Requests? 
From: RFC 9110:  https://www.rfc-editor.org/rfc/rfc9110.html#section-11.6.2
>The "Authorization" header field allows a user agent to authenticate itself with an origin server -- usually, but not necessarily, after receiving a [401 (Unauthorized)](https://www.rfc-editor.org/rfc/rfc9110.html#status.401) response.

Reply:
(assuming you mean with HTTP Responses) You're right. I'm not sure where I got the notion it's used by both requests and response.

I'll include this in the Authorization field.